### ISSUE TITLE
Invalid schema after Quarkus update

### ISSUE THREAD

**Issue by KMUS** [leangen/graphql-spqr#394]
The GraphQL schema can no longer be injected after the Quarkus version has been updated from 1.12.2 to 1.13.x. This only happens when Quarkus is running in development mode. Quarkus version 1.12.2 is the last one that works fine with SPQR schema injection.
I am grateful for any help. 

I am using the last SPQR version.

```
java.lang.RuntimeException: Error injecting graphql.schema.GraphQLSchema x.y.z.rest.resource.GqlHttpEndpoint.schema
	at x.y.z.rest.resource.GqlHttpEndpoint_Bean.create(GqlHttpEndpoint_Bean.zig:379)
	at x.y.z.rest.resource.GqlHttpEndpoint_Bean.create(GqlHttpEndpoint_Bean.zig:445)
	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:96)
	at io.quarkus.arc.impl.AbstractSharedContext.access$000(AbstractSharedContext.java:14)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:29)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:26)
	at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:26)
	at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:26)
	at x.y.z.rest.resource.GqlHttpEndpoint_Bean.get(GqlHttpEndpoint_Bean.zig:477)
	at x.y.z.rest.resource.GqlHttpEndpoint_Bean.get(GqlHttpEndpoint_Bean.zig:493)
	at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:430)
	at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:443)
	at io.quarkus.arc.impl.ArcContainerImpl$1.get(ArcContainerImpl.java:266)
	at io.quarkus.arc.impl.ArcContainerImpl$1.get(ArcContainerImpl.java:263)
	at io.quarkus.arc.runtime.BeanContainerImpl$1.create(BeanContainerImpl.java:35)
	at io.quarkus.resteasy.common.runtime.QuarkusConstructorInjector.construct(QuarkusConstructorInjector.java:54)
	at org.jboss.resteasy.plugins.server.resourcefactory.POJOResourceFactory.createResource(POJOResourceFactory.java:71)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:386)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:68)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at org.jboss.resteasy.plugins.server.servlet.ServletContainerDispatcher.service(ServletContainerDispatcher.java:249)
	at io.quarkus.resteasy.runtime.ResteasyFilter$ResteasyResponseWrapper.service(ResteasyFilter.java:70)
	at io.quarkus.resteasy.runtime.ResteasyFilter$ResteasyResponseWrapper.sendError(ResteasyFilter.java:76)
	at io.undertow.servlet.handlers.DefaultServlet.doGet(DefaultServlet.java:172)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:503)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:590)
	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:74)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129)
	at io.quarkus.resteasy.runtime.ResteasyFilter.doFilter(ResteasyFilter.java:31)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.quarkus.micrometer.runtime.binder.vertx.VertxMeterBinderUndertowServletFilter.doFilter(VertxMeterBinderUndertowServletFilter.java:28)
	at javax.servlet.http.HttpFilter.doFilter(HttpFilter.java:97)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:63)
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at io.undertow.servlet.handlers.RedirectDirHandler.handleRequest(RedirectDirHandler.java:67)
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:133)
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:65)
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.security.handlers.NotificationReceiverHandler.handleRequest(NotificationReceiverHandler.java:50)
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:247)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:56)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:111)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:108)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at io.quarkus.undertow.runtime.UndertowDeploymentRecorder$9$1.call(UndertowDeploymentRecorder.java:587)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:227)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleRequest(ServletInitialHandler.java:152)
	at io.quarkus.undertow.runtime.UndertowDeploymentRecorder$1.handleRequest(UndertowDeploymentRecorder.java:119)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:290)
	at io.undertow.server.DefaultExchangeHandler.handle(DefaultExchangeHandler.java:18)
	at io.quarkus.undertow.runtime.UndertowDeploymentRecorder$5$1.run(UndertowDeploymentRecorder.java:413)
	at io.quarkus.runtime.CleanableExecutor$CleaningRunnable.run(CleanableExecutor.java:231)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2415)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at java.base/java.lang.Thread.run(Thread.java:834)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
Caused by: graphql.schema.validation.InvalidSchemaException: invalid schema:
"Query" must define one or more fields.
	at graphql.schema.GraphQLSchema$Builder.buildImpl(GraphQLSchema.java:647)
	at graphql.schema.GraphQLSchema$Builder.build(GraphQLSchema.java:622)
	at io.leangen.graphql.GraphQLSchemaGenerator.generate(GraphQLSchemaGenerator.java:1036)
	at x.y.z.graphql.producer.GraphQLSchemaProducer.getSchema(GraphQLSchemaProducer.java:73)
	at x.y.z.graphql.producer.GraphQLSchemaProducer_Subclass.getSchema$$superaccessor1(GraphQLSchemaProducer_Subclass.zig:189)
	at x.y.z.graphql.producer.GraphQLSchemaProducer_Subclass$$function$$11.apply(GraphQLSchemaProducer_Subclass$$function$$11.zig:29)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:54)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.proceed(InvocationInterceptor.java:63)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.monitor(InvocationInterceptor.java:49)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor_Bean.intercept(InvocationInterceptor_Bean.zig:521)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
	at x.y.z.graphql.producer.GraphQLSchemaProducer_Subclass.getSchema(GraphQLSchemaProducer_Subclass.zig:147)
	at x.y.z.graphql.producer.GraphQLSchemaProducer_ProducerMethod_getSchema_c237c3d1d90fada122f993db962dc05fcc1c4a18_Bean.create(GraphQLSchemaProducer_ProducerMethod_getSchema_c237c3d1d90fada122f993db962dc05fcc1c4a18_Bean.zig:130)
	at x.y.z.graphql.producer.GraphQLSchemaProducer_ProducerMethod_getSchema_c237c3d1d90fada122f993db962dc05fcc1c4a18_Bean.get(GraphQLSchemaProducer_ProducerMethod_getSchema_c237c3d1d90fada122f993db962dc05fcc1c4a18_Bean.zig:160)
	at x.y.z.graphql.producer.GraphQLSchemaProducer_ProducerMethod_getSchema_c237c3d1d90fada122f993db962dc05fcc1c4a18_Bean.get(GraphQLSchemaProducer_ProducerMethod_getSchema_c237c3d1d90fada122f993db962dc05fcc1c4a18_Bean.zig:195)
	at x.y.z.rest.resource.GqlHttpEndpoint_Bean.create(GqlHttpEndpoint_Bean.zig:362)
	... 78 more
```

**Comment by kaqqao** [comment]
Any chance you figured this out?
If not, could you please provide a minimal example that replicates the problem, so I can investigate?
The exception itself only says no fields were found on a type. Since you say this only happens in development and only on new Quarkus versions, I'm quite confident it has to do with class loading. Here's my theory. SPQR is looking for annotations on a class. The class gets reloaded by a different class loader in development mode, and so do the annotation classes themselves. SPQR still holds references to the original annotation classes and looks for them on the reloaded classes. The new and old annotation classes don't match, so SPQR finds no matches, extracts no fields and you get the exception.

You could test this theory by adding a breakpoint in `AnnotatedResolverBuilder#isQuery` and see if the two `@GraphQLQuery` classes match.

If I'm right, you'd have to somehow re-execute the schema generation when query classes are reloaded... But I really don't know enough about Quarkus to say more than that.

**Comment by KMUS** [comment]
I am sorry for the long delay. Today, I refactored the code, so I can provice a minimal example. It is still the same behaviour, Quarkus version 1.12.2 works well, 1.13.7 gives invalid schema.

What I notified, the working version calls the method of ProjectProfileGql once, the broken one twice in AnnotatedResolverBuilder#isQuery(Method method, ResolverBuilderParams params).

[graph.zip](https://github.com/leangen/graphql-spqr/files/6741599/graph.zip)

**Comment by kaqqao** [comment]
Great, thanks for a full example! I'll try debugging to see what I can find...

**Comment by kaqqao** [comment]
Alright, I found the issue. It's simple to deal with, but the reasoning behind it is complex.

#### TL;DR
Register the bean with an explicit type: `.withOperationsFromSingleton(services, ProjectProfileGql.class)` and it will work.

#### The long explanation:
You registered a bean like this:
```java
.withOperationsFromSingleton(services)
```
but this is problematic for 2 reasons:

1) `services` isn't a singleton 😵 It's `@Dependent`!
2) Quarkus enhances the `ProjectProfileGql` class, so the real type of the injected `services` is `ProjectProfileGql_Subclass`, and this class contains no annotations, so SPQR can't find them.

Issue 1 isn't really an issue in this case, since the `services` instance is actually stateless, so it might as well be treated as a singleton. But be careful around registering non-singleton instances this way because if they ever do become stateful, you'll get strange behaviors. In general, dealing with non-singletons is complex enough that I'd recommend staying far away from them.

Issue 2 is much more important in this case. In Java, method-level annotations are not  inherited, and there's no surefire way of even telling which method overrides which. Because Spring, Quarkus and other frameworks wrap, instrument and subclass types dynamically to provide various features, they need a very complicated logic for discovering what annotation applies where. SPQR doesn't have this logic, or at least not generically (the Spring starter has some Spring-specific logic), as it might even differ based on the manipulation the framework in question performed, so you have to help it in detecting the actual type of the instance where the annotations are. That's why giving the class explicitly when registering the bean solves this.

I'll close the issue, but feel free to reopen if you need more help.

**Comment by kaqqao** [comment]
Note: I edited the previous comment quite a lot, so if you're reading the original via email, please make sure to check out the updated version on the web.

**Comment by KMUS** [comment]
@kaqqao 
Thank you for your time and figguring this out. But how can I solve it with multiple classes? AbstractGqlEndpoint is our super class and every GraphQL class inherits from it.

```
@Inject
List<AbstractGqlEndpoint> services;

.withOperationsFromSingletons(services.toArray(new Object[] {}))
```

Does it mean, I have to include evey class instance in withOperationsFromSingletons like this?

`.withOperationsFromSingletons(services.toArray(new A(), new B(), ...)`

**Comment by kaqqao** [comment]
Hmm... I'm not sure what's the best way as I'm not familiar with Quarkus... E.g. Spring gives you tools to detect the user-defined type from a managed bean. If such a thing does not exist, you could perhaps make a workaround yourself:

```java
//Typed in the browser, so not sure if it's exactly right
for (AbstractGqlEndpoint service : services) {
    generator.withOperationsFromSingleton(service, unwrapClass(service.getClass));
}

private Class<?> unwrapClass(Class<?> clazz) {
    if (clazz.getName().endsWith("_Subclass")) {
	  return Class.forName(clazz.getName().replace("_Subclass", ""), true, clazz.getClassLoader());
    }
    return clazz;
}
```

I opened a new issue #402 to address this systematically in SPQR. So I hope workarounds such as the above will only be needed briefly.

**Comment by kaqqao** [comment]
Actually... This is likely enough in your case:

```java
private Class<?> unwrapClass(Class<?> clazz) {
    if (clazz.getName().endsWith("_Subclass")) {
        return class.getSuperClass();
    }
    return clazz;
}
```
But, again, this is just a temporary work around. I'm working on a more appropriate solution (#402).

**Comment by KMUS** [comment]
Many thanks for your valueable help. That little hack seems to fix the issue. Quarkus 2.0.3.final and SPQR 0.11.2. run fine now.
